### ISSUE TITLE
IMAP login using an alias works unexpectedly

### ISSUE THREAD

**Issue by mutech** [docker-mailserver/docker-mailserver#3150]
# Subject

Other

## Description

I migrated an outdated tvial/mailserver to docker-mailserver (actually only the postfix-{accounts, virtual}.cf files).

I accidentally used an alias (given.family@domain instead of gfamily@domain) as account name configuring the IMAP client and the email client did not complain successfully connected to dovecot. I did not see incoming mails in that mailbox, other than copies of testmails I sent using the same credentials on that client using the postfix of that mailserver instance.

Sending email using the alias/password also works (that's actually a nice feature).

**postfix-accounts.cf**
```TEXT
gfamily@domain|{SHA512}...
```

**postfix-vrtual.cf**
```TEXT
given.family@domain gfamily@domain
```

Mails to both `gfamily` and `given.family` get properly dispatched to the account `gfamily@domain`.

## Expected Behavior

I expected a login using an alias name and the correct password of the target user of that alias to fail.

Alternatively, it would be cool if the alias would be silently resolved to a target account with a matching password ("a matching", because an alias could refer to multiple targets), but then it would have had to be connected to the correct dovecot account.

As it is, I now have a stale dovecot account with emails in the sent folder that I would have to migrate to the correct account if it were real mails. No problem in this case, because it's my own account. But if I would be administering a corporate server, I would get cold feet and stop planning weekend activities.

## Remarks

It's perfectly possible that this is a consequence of me not using `setup.sh` and instead just copying the old `.cf` files. But I expect that to work, because if I'm restoring a backup I also won't call setup.sh. Then the dovecot accounts would most likely also be restored, but that's not necessarily the case. As far as I can tell, for format or semantics of these files did not change for decades.

**Comment by mutech** [comment]
@georglauterbach how should I interpret the assignment?

My use case for mailserver changed, so that I am probably not using this setup anymore: I want to make use of dovecot's bidirectional synchronization, keeping one instance on premise and one online. Consequently I should probably splt mta from imap, have two mta's and two dovecots but not using the same docker hosts; in that scenario docker-mailserver is probably not the right choice.

In this circumstance, the problem I was reporting solves itself for me, so if I'm assigned my preferred action would be to close the issue.

On the other hand I reported the issue to save people from having to find the cause of this behavior, because it's actually difficult to find it if what you have is a user reporting that they don't receive email while they can successfully connect to the server. The only reason why it took me only hours to find what's wrong is because I was testing the server after a new installation.

I'm closing it, feel free to reopen if you think this is relevant or query if you need more information about my setup, e.g. if you can't reproduce it.

**Comment by georglauterbach** [comment]
I assigned you because I do this by default. The issue is that we add the aliases to the Dovecot user accounts because our alias code is probably just bad, and we had an issue raised complaining about warnings/errors in the logs when delivering mail to aliases.

If you know how to properly configure aliases with Postfix and Dovecot, please help us :)

**Comment by mutech** [comment]
> The issue is that we add the aliases to the Dovecot user accounts because our alias code is probably just bad, and we had an issue raised complaining about warnings/errors in the logs when delivering mail to aliases.
> 
> If you know how to properly configure aliases with Postfix and Dovecot, please help us :)

I'm using the precursor of this project (if the tvial-repository is its precursor) since so long that I don't even remember when I didn't. I never had to touch anything, that's why I realized so late that the project moved, I just literally forgot that a mailserver needs maintenance every now and then :-)

So I really would not say that your code is bad. 

Btw., the updated from my archaeologically significant instance to the current version is a massive improvement in terms of quality of out-of-the-box spam recognition and overall performance. I feel.... modern... in a way :)

But yes, of course I would like to help and I guess I could fix this in terms of understanding. My problem is that I also only have limited time and my use case for self-hosting mail changed over the years. The monolithic nature of docker-mailserver makes it hard for me to adopt it to my needs. I'm going to describe it here, going off-topic on my own issue - *grin* - to see if my direction is attractive to you guys. If so, this will touch on authentication anyway and probably solve both of our issues as a side effect.

I'm using my email server to send and receive email, but also as a cloud storage space and archive for licenses, invoices, document management system and search engine and all kinds of stuff that accumulates by email. That's not a strategy, but of all technologies I was and am using, email is the one that never failed, the normative power of the factual or Darwin at work.

This is of course deeply inadequate. Data on the internet is not encrypted, my backup strategy is incomplete to say the least and I have a mail client on each device in my homelab, each tablet, phone or laptop I'm using and they are all polling the external email server eating up to a third of my network bandwidth. It's ridiculous. But however crazy this setup is, it's hard to argue with success, ask the next dinosaur.

Looking for a compromise between Darwin and Reason, I found out that dovecot supports both encryption at rest (using the login password as key) and bidirectional synchronization, I saw the light. Having one instance in the cloud and one on premise solves the bandwidth problem (letting the mesh of email clients poll locally) and the availability issue (always having all data accessible) as well as the privacy problem of insecure cloud storage. I don't need redundancy in the cloud for dovecot, because mta's will keep pending messages long enough for me to setup a new instance via ansible and synchronization keeps track of data recovery.

Problem with this approach is that dovecot can only handle two instances using replication this way, and other general imap replication mechanisms I found seem to be less robust than dovecots approach.

docker-mailserver is tightly integrated. It's probably possible to link up dovecot in mailserver with another instance manually but that does not seem to be a good approach.

I think the best approach would be to split docker-mailserver in modules, one mta container, one imap container, later maybe a webmail container and an admin UI container. That would of course create a strong incentive to also define another module containing the user/account configuration. That's probably a good thing, because while the current approach using postfix .cf files is sexy, many will refrain from using mailserver because they need to interface whatever directory or database they are using for their accounts (ldap/AD/...). That way, this issue just goes away (and yes I know that this issue is nothing in comparison to the work resulting from my concept).

I'm reluctant to start working on this because I will almost certainly not have the time to maintain it. It would be cool to do it together with you guys, but that only works if you like the approach and see a benefit. What do you think, does this sound attractive?

**Comment by mutech** [comment]
> If you know how to properly configure aliases with Postfix and Dovecot, please help us :)

The only problem I remember ever having with alias configurations was in conjunction with catchall/wildcard setups. There the target account did not get served because aliases have a higher priority than accounts (as I remember it), thus to make it work, I had to alias the account to itself (`@domain user@domain; user@domain user@domain`). This is from the top of my head, forgot the details...

Other than that, I never had any issues with aliases. What were the reported problems about?

Thinking again about it, being able to log on with username = alias and the password from one of the target accounts is actually a really cool feature, but only if you actually would get access to the data, which now doesn't work. I don't think this is a sustainable workaround for whatever the original problem was.

**Comment by georglauterbach** [comment]
Yeah, this is the thing we do have serious issues with I think: catch-all aliases. Moreover, you see that aliases are added Dovecot's userdb, but probably not properly.. and we're always looking for maintainers - so if you want to chime in and look where the problem of this issue is, that would be appreciated :D

**Comment by polarathene** [comment]
> Sending email using the alias/password also works (that's actually a nice feature).

It's also possibly just from a workaround. A workaround that isn't ideal but removing it requires either disabling Dovecot Quotas feature (which we support), or writing our own policy service proxy for Postfix to communicate with Dovecot (_Postfix doesn't resolve the alias at this point IIRC, and just checks with Dovecot if there is an account for it that is over quota limit, which it's not without the workaround of creating dummy accounts_).

That workaround only supports basic aliases, if the alias is for multiple accounts and/or aliases, it'll fail as there is no direct 1:1 mapping from an alias to a single mail account.

---

> I expected a login using an alias name and the correct password of the target user of that alias to fail.

Disable the Dovecot Quotas feature if that's acceptable.

> As far as I can tell, for format or semantics of these files did not change for decades.

`docker-mailserver` itself has only been around since 2015-2016 IIRC. If you backup anything that is config from us (_`postfix-accounts.cf` and the like_), these AFAIK are using a custom format, not one that we can directly pass to Postfix or Dovecot.

As such, be mindful that major releases may break compatibility. Generally that's not been the case, but it is possible in future that these configs change in format and filename / location. Parsing of them in our start-up scripts and `setup.sh` utilities has more churn though, where compatibility is more likely affected.

---

> I think the best approach would be to split docker-mailserver in modules, one mta container, one imap container, later maybe a webmail container and an admin UI container.

AFAIK you can presently configure the container to just focus on being an MTA (disabled Dovecot) with `SMTP_ONLY=1`, but we do not have anything that disables Postfix, and only runs Dovecot. That could be approved if it's not too invasive of a change to maintain (eg: `ENABLE_POSTFIX=0`).

Alternatively if you just need Dovecot, and can get away with a companion image providing just that (_there's an [official Dovecot image](https://hub.docker.com/r/dovecot/dovecot/)_) it may be the better way to go with, and if you're up for it, adding an example in our docs for other users to know how to configure a setup like that.

For webmail, you can do this with other Docker images like for RainLoop or any other preferred webmail project, I'm not sure if it's something we'd integrate.

---

> That would of course create a strong incentive to also define another module containing the user/account configuration.

There has been requests for an Admin UI, and we have a separate project in our org for this that AFAIK has gone stale. It's goal was to define an OpenAPI schema and provide an API + web UI I think to at least allow you to change account passwords via a web UI. It would use it's own `Dockerfile` that extended our releases so that you could deploy a single image.

My personal preference would be a separate project that handles the API integration with our `setup` command, and this could be included in the official image to be enabled via ENV. Use a unix or tcp socket to communicate between the two containers, and that other container provides the web UI, or exposes a client-side API that others can build / integrate their own web UIs on top of.

I spent a fair amount of time cleaning up the account management support via `setup`, but it's not quite complete, especially for any traffic with high frequency of changes (larger scale deployment) we're not suitable for handling.

Our internal local account management scripts and support is potentially a candidate to be extracted out to a separate project that `docker-mailserver` imports in the `Dockerfile` build and integrates with. This would likely involve a port to a rust binary and a change in the config file format, but is quite demanding of time required to do that I'm not sure if we'll ever see it happen tbh :sweat_smile: 

**NOTE:** We support the default local file configs for user accounts, in addition to LDAP, and there has been talk about OIDC/OAUTH as another addition.

Using files instead of a DB seems convenient but has a variety of maintenance and scaling issues, especially when you have our change-detection support in mind and need to consider multiple reader/writers on a config file which requires properly implementing file locking (_which regressed in the 10.x series because someone got changes approved to change it for NFS support, but likely had NFS misconfigured, nor does some services support NFS properly such as Dovecot under configs we ship_).

If our LDAP support wasn't in bad shape, it would probably be a better choice. Delegating user account management to an external provider via OIDC/OAUTH provider would be another potentially better alternative, but may not integrate any smoother than LDAP for some feature support.

> That's probably a good thing, because while the current approach using postfix .cf files is sexy, many will refrain from using mailserver because they need to interface whatever directory or database they are using for their accounts (ldap/AD/...). That way, this issue just goes away (and yes I know that this issue is nothing in comparison to the work resulting from my concept).

Perhaps I didn't follow what you had in mind well?

Our LDAP support tries to delegate away user config from our responsibility but still needs integration on our end. You get the benefit of managing accounts via LDAP such as some UI, but you'll see that the support is problematic / decaying over time since that contributor of the feature doesn't maintain it. 

Our users generally have less bugs without LDAP, or at least the LDAP ones are not easy to investigate / resolve for us (_see the stale draft PR regarding `/var/mail` ownership / permissions_).

---

> I'm reluctant to start working on this because I will almost certainly not have the time to maintain it.

I'd suggest starting with a community contribution to the docs. It's less burden on maintainers, where we've struggled to attract more help.

For me personally I want to move on from maintaining the project (_I've never actually deployed DMS since contributing improvements from late 2020_), it's difficult for me to continue justifying the toll it costs me in not only time to be as active on the project as I like to :sweat_smile: (_I love being a maintainer here and working on the project, it's just not sustainable for me_)

**Comment by mutech** [comment]
> Yeah, this is the thing we do have serious issues with I think: catch-all aliases.

Well if it's that, then it might really just be that issue with aliases shadowing the account. That would just need an update of the documentation if it can be verified.

> Moreover, you see that aliases are added Dovecot's userdb, but probably not properly.. and we're always looking for maintainers - so if you want to chime in and look where the problem of this issue is, that would be appreciated :D

Well if I stay with docker-mailserver, I would gladly join you guys. But I first have a bucket list to work on before I can get back to this topic. I'm currently drowning in ansible scripting, reworking all of my infrastructure and the same on the job. When I have some air to breath I'm checking back on this.

**Comment by mutech** [comment]
We're getting into the realm where we might want to have a live chat. It's already getting difficult to keep track of the threads ;-)

> > Sending email using the alias/password also works (that's actually a nice feature).
> It's also possibly just from a workaround. A workaround that isn't ideal but removing it requires either disabling Dovecot Quotas feature (which we support), or writing our own policy service proxy for Postfix to communicate with Dovecot (_Postfix doesn't resolve the alias at this point IIRC, and just checks with Dovecot if there is an account for it that is over quota limit, which it's not without the workaround of creating dummy accounts_).

I think the biggest problem with the workaround is that it's actually not working. That doesn't matter as long as there are no users stumbling over that issue (which is unlikely), but when they do it will be very painful for the guy in charge of fixing things.

> > I expected a login using an alias name and the correct password of the target user of that alias to fail.
> 
> Disable the Dovecot Quotas feature if that's acceptable.

In my case that's fine. Now that I know about it, I can just ignore it or disable quotas, both works for me.

> Alternatively if you just need Dovecot, and can get away with a companion image providing just that (_there's an [official Dovecot image](https://hub.docker.com/r/dovecot/dovecot/)_) it may be the better way to go with, and if you're up for it, adding an example in our docs for other users to know how to configure a setup like that.

The sweet spot of your project is that it's maintained as a whole. You did that so well that I really could just ignore the topic mail delivery for years. That's really why I'm so reluctant to change.

> For webmail, you can do this with other Docker images like for RainLoop or any other preferred webmail project, I'm not sure if it's something we'd integrate.

I mentioned that because a lot of people need the feature. I myself don't like or use webmail at all.

> > That would of course create a strong incentive to also define another module containing the user/account configuration.
> 
> There has been requests for an Admin UI, and we have a separate project in our org for this that AFAIK has gone stale. It's goal was to define an OpenAPI schema and provide an API + web UI I think to at least allow you to change account passwords via a web UI. It would use it's own `Dockerfile` that extended our releases so that you could deploy a single image.

I guess most people don't want to administer their email accounts separately and instead just maintain their active directory, LDAP or FreeIPA or people like me where email settings rarely if ever change. I'm personally perfectly happy with these two `.cf` files. Those who are looking for a web interface will probably be running from the task of using a setup script or editing the postfix files. But they will probably not be your user base either way.

> My personal preference would be a separate project that handles the API integration with our `setup` command, and this could be included in the official image to be enabled via ENV. Use a unix or tcp socket to communicate between the two containers, and that other container provides the web UI, or exposes a client-side API that others can build / integrate their own web UIs on top of.

I'm currently doing a lot of ansible scripting. From that perspective the config files are easily digestable for me, much smoother than an API, since I already have the issue with authentication and getting onto the server solved there and updating files using templates and data from the inventory is the easiest way.

I think the API is then most relevant if you actually have to support more heterogeneous settings, such as separate hosting of MTA and IMAP or integration of LDAP and stuff. And then it's probably best to use these integrations, since they are already there and well established standards, even if ugly and uncomfortable to use.

> I spent a fair amount of time cleaning up the account management support via `setup`, but it's not quite complete, especially for any traffic with high frequency of changes (larger scale deployment) we're not suitable for handling.

Yeah, that's not the strength of this project. Doesn't have to be either I guess.

> Our internal local account management scripts and support is potentially a candidate to be extracted out to a separate project that `docker-mailserver` imports in the `Dockerfile` build and integrates with. This would likely involve a port to a rust binary and a change in the config file format, but is quite demanding of time required to do that I'm not sure if we'll ever see it happen tbh sweat_smile

> **NOTE:** We support the default local file configs for user accounts, in addition to LDAP, and there has been talk about OIDC/OAUTH as another addition.

Well if you support ldap already, then you were either hiding it well or I have a serious blind spot. I am not aware of that at all. :)

> Using files instead of a DB seems convenient but has a variety of maintenance and scaling issues, especially when you have our change-detection support in mind and need to consider multiple reader/writers on a config file which requires properly implementing file locking (_which regressed in the 10.x series because someone got changes approved to change it for NFS support, but likely had NFS misconfigured, nor does some services support NFS properly such as Dovecot under configs we ship_).

I think the file based approach really only makes sense in the single instance configuration I know mailserver. In any other scenario it's doomed to inflict pain. I understood the change detection as a pure update feature and never expected mailserver to cope with concurrency of any kind.

> If our LDAP support wasn't in bad shape, it would probably be a better choice. Delegating user account management to an external provider via OIDC/OAUTH provider would be another potentially better alternative, but may not integrate any smoother than LDAP for some feature support.

It really depends. For my old setup, LDAP would just be another nuisance. In a corporate context or when you split MTA from IMAP as I'm planning, things change and the file based setup changes from a feature to debts. But the question is really that of the scope of this project. You do simple setup extremely well. 

> > That's probably a good thing, because while the current approach using postfix .cf files is sexy, many will refrain from using mailserver because they need to interface whatever directory or database they are using for their accounts (ldap/AD/...). That way, this issue just goes away (and yes I know that this issue is nothing in comparison to the work resulting from my concept).
> 
> Perhaps I didn't follow what you had in mind well?

That's what I tried to explain above. File setup working well for single all-in-one settings and not so well in a more decoupled or more corporate environment. There LDAP and the like are probably just working better.

> Our LDAP support tries to delegate away user config from our responsibility but still needs integration on our end. You get the benefit of managing accounts via LDAP such as some UI, but you'll see that the support is problematic / decaying over time since that contributor of the feature doesn't maintain it.

Ya, I can confirm, working with LDAP is not fun. Nobody really wants to maintain that stuff. Don't look at me btw. No way I'm getting myself into that stuff :)

> Our users generally have less bugs without LDAP, or at least the LDAP ones are not easy to investigate / resolve for us (_see the stale draft PR regarding `/var/mail` ownership / permissions_).
> 
> > I'm reluctant to start working on this because I will almost certainly not have the time to maintain it.
> 
> I'd suggest starting with a community contribution to the docs. It's less burden on maintainers, where we've struggled to attract more help.

> For me personally I want to move on from maintaining the project (_I've never actually deployed DMS since contributing improvements from late 2020_), it's difficult for me to continue justifying the toll it costs me in not only time to be as active on the project as I like to sweat_smile (_I love being a maintainer here and working on the project, it's just not sustainable for me_)

That's sad. It sounds like you guys are working well together and the result is really nice.

**Comment by georglauterbach** [comment]
> When I have some air to breath I'm checking back on this.

Sounds great! We appreciate every help we can get 🙂

**Comment by polarathene** [comment]
> We're getting into the realm where we might want to have a live chat. It's already getting difficult to keep track of the threads ;-)

Yeah sorry about that, I can be quite verbose (_takes too long for me to slim it down_). No obligation to respond to everything :+1: 

---

> I think the biggest problem with the workaround is that it's actually not working.

The workaround is working, it is for Dovecot Quota feature to prevent scammers exploiting you for Backscatter (bad).

The login as an alias is a side-effect that wasn't intended.

---

> I mentioned that because a lot of people need the feature. I myself don't like or use webmail at all.

My plan was to just get around to adding a docs page for integration, and possibly referencing an example config in the repo.

There is a few gotchas involved with that IIRC due to fail2ban for example seeing all login traffic from the webmail service as a single IP without the admin configuring remote IP header forwarding or something like that. It exists in our issue history, but the users never contributed to docs for the benefit of others.

> I'm currently doing a lot of ansible scripting. From that perspective the config files are easily digestable for me, much smoother than an API

Config files approach we have for accounts isn't going away, nor would the `setup` CLI tool. The API would just be a way to cater to needs of those wanting to provide user-facing UI to manage accounts (_eg: integration into a larger products frontend that manages more than DMS_).

---

> Well if you support ldap already, then you were either hiding it well or I have a serious blind spot. I am not aware of that at all. :)

It's purely community contributed from years ago and continues to receive reports of problems that no active maintainer is able to really resolve.

We have a few docs pages, but from what I've heard LDAP providers vary, and the example configs aren't necessarily reliable. I think we may even have a mix in the docs where the demonstrated configs aren't interoperable / consistent.

> File setup working well for single all-in-one settings and not so well in a more decoupled or more corporate environment. There LDAP and the like are probably just working better.

Right, and generally corporate environments will find other projects better suited to them, or paid solutions. I'm not familiar with the type of deployments our LDAP users are doing, just that there's been a fair amount chiming in with issues (_many not resolving them or contributing back, usually switching to another solution instead of DMS_).

LDAP is apparently a feature we offer that other popular alternatives are not :man_shrugging: 

---

> That's sad. It sounds like you guys are working well together and the result is really nice.

We're a great and friendly team! :grinning: 

It's just an issue with cost of time invested due to our limited size of active maintainers / contributors. I have spent the past two months almost full time on DMS when I really should not have, and still have a huge backlog of things I'd love to tackle for DMS. The upcoming v12 fixes a bunch of issues and almost got our test suite to reach a milestone I've been wanting to reach for some time :tada: 

As I don't rely on DMS (_although I had intended to deploy it to assist another community I help_), I can't sustain the effort when I need to be focused on seeking paid work :sweat_smile: 

I care about DMS doing well and will continue to contribute where possible, but need to reduce the responsibility I've imposed on myself. I was hoping we'd have attracted more maintainers over time (_we seemed to go into a deficit instead_ :stuck_out_tongue: ) The other maintainers are very capable though and I learned heaps collaborating with them :sunglasses:
